Add interoperable module guide (#1903)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/build-blockchain/interoperable-module/index.adoc

== Testing interoperable modules

As interaction between two sidechains always involves three chain (sending chain, receiving chain, and mainchain), it is quite complicated to test newly implemented interoperable modules locally.

To simplify the process, the `lisk-sdk` provides useful example scripts that can be easily adjusted to try out new interoperable modules.

=== The `interop` example of the `lisk-sdk` repository

[.float-group]
--
[role="right"]
.The blockchain network of the interop example
image::build-blockchain/interop-example.png[,300,role="right"]

To try out the cross-chain communication of interoperable modules locally, go the {url_github_sdk_interop}[interop/example] in the `lisk-sdk` repository.

Here, you find a `README.md` which is describing all required steps to set up a simple interoperability example, as illustrated in _Figure 1_.

Additionally, you find examples how to execute cross-chaon token transfers between the chains.

Follow the instructions until {url_github_sdk_interop_cctransfer}[here^], where you successfully sent tokens from the mainchain to both sidechains.

Verify the sidechain account is funded with mainchain tokens::

> The account sending the `reactCrossChain` command, whichwill create a new CCM needs to provide the fees for the execution of the CCM in tokens that are native on the mainchain.

Author: ```suggestion
The account sending the `reactCrossChain` command, which will create a new CCM, needs to provide the fees for the execution of the CCM in tokens that are native on the mainchain.
```

> This is why it should be verified, that the account that intends to send the transaction own enough mainchain tokens.
> If you followed the instructions in the `README.md` until {url_github_sdk_interop_cctransfer}[here^], the account `lskx5uqu2zzybdwrqswd8c6b5v5aj77yytn4k6mv6` of `pos-sidechain-exmaple-two` should have the required funding.

Author: ```suggestion
This is why it should be verified, that the account, that intends to send the transaction, owns enough mainchain tokens.
If you followed the instructions in the `README.md` until the {url_github_sdk_interop_cctransfer}[Transfer from mainchain to sidechain two^] step, the account `lskx5uqu2zzybdwrqswd8c6b5v5aj77yytn4k6mv6` of `pos-sidechain-exmaple-two` should have the required funding.
```

}
{
  "transaction": {
    "module": "react",
    "command": "reactCrossChain",
    "fee": "1000000",
    "nonce": "0",
    "senderPublicKey": "c94952af78216ff92d615f4eb566726693d05df2d0b82f7fce26fc1f0e6e8724",
    "signatures": [
      "4f68f83f8e32aadba4b2bb465ac0b09ca6a1c0ca247750e21d22cbd0fc53a5684fc3d4185d47e69e6e85a818ef04a61430dcadc096813cb205f735d04e3c4901"
    ],
    "params": {
      "reactionType": 0,
      "helloMessageID": "lskmjwp8z88avvxn4voktmx8cu9wk4opjkna5owt5",
      "receivingChainID": "04000001",
      "data": "",
      "messageFee": "2100000"
    },
    "id": "5043174c97f508b711d9be0ec57ed60009ea83b57b2a665cef8c99420b9fcbb2"
  }
}
Transaction with id: '5043174c97f508b711d9be0ec57ed60009ea83b57b2a665cef8c99420b9fcbb2' received by node.
----
=== Observe execution of the CCM

Sidechain 2::

> * http://localhost:4007[^]

Author: yes

    "status": 0
  }
}
----

Next, go to the events of the Dashboard of *sidechain 1* and look for the following event:

 interoperability_ccmProcessed (height: 335 index: 3 topics: ab36b23bce5e2a0cebe9131c5b3bddfd8132e9367f1cf9a95875ae74b8ca3909, f092606d704d8a205bf2d702119c4761e3cb8cc22197fd28411f28fae9aa4d98, 04000002, 04000001)

.Dashboard sidechain 1
[source,json]
----
{
  "ccm": {
    "module": "hello",
    "crossChainCommand": "crossChainReact",
    "nonce": "1",
    "fee": "100000000",
    "sendingChainID": "04000002",
    "receivingChainID": "04000001",
    "params": "080112296c736b367134747a79657633386666776d66756263746163717a6f667962377a326f667266666d33791a0772656163746564",
    "status": 0
  },
  "result": 0,
  "code": 0
}
----

If there is an event similar to the above example, it means that the CCM from sidechain 2 was successfully processed on sidechain 1.

Once the CCM is received and executed on the node, it is possible to request the reaction on sidechain 1 with the endpoint `hello_getReactions` which we implement in the section xref:{url_build_executeccm_endoint}[Creating a new endpoint for getting reactions for a Hello message].

.Sidechain 1:
[source,bash]
----

> ./bin/run endpoint:invoke hello_getReactions '{"address":"lskmjwp8z88avvxn4voktmx8cu9wk4opjkna5owt5"}

Reviewer: ```suggestion
./bin/run endpoint:invoke hello_getReactions '{"address":"lskmjwp8z88avvxn4voktmx8cu9wk4opjkna5owt5"}'
```

----

.Response, after one account has liked the message
----
{"reactions":{"likes":[{"type":"Buffer","data":[128,101,154,216,166,239,226,225,198,252,61,66,236,115,108,201,97,49,120,140]}]}}
----

By receiving this response from the blockchain store of sidechain 1, it is verified that the cross-chain reaction to Hello messages was implemented successfully.

=== Tips for debugging the interoperability example

==== How to get the last sent CCM

To get the last sent CCM on a chain, invoke the `chainConnector_getLastSentCCM` endpoint.

TIP: Remember to adjust the value `--data-path` to the correct path for the respective node.
The command can be run from any node CLI, and will always respond with data from the node specified in the `data-path`.

.pos-sidechain-example-two
[source,bash]
----

> ./bin/run endpoint:invoke chainConnector_getLastSentCCM --pretty --data-path=/Users/mona/.lisk/mainchain-node-one

Reviewer: Please remove the `datapath` property, it throws an error when executed exactly since the user's name might differ. 

I have tried it without the `datapath` and it works fine. I would also recommend removing the datapath from the next few snippets.
```suggestion
./bin/run endpoint:invoke chainConnector_getLastSentCCM --pretty
```

Author: replaced it with a more general path

----

> ==== How to authorize the chain connector plugin

Reviewer: I missed performing this step earlier, and I think that is why the example didn't work for me. Is it possible that this step could be moved before the `Create Reaction on sidechain 2`? Or are these two things not related to each other.

Author: this only needs to be done when restarting a node, the authorization is already part of the README as well as the start example script, so it shouldn't be missed

In case it is necessary to restart the nodes, it is necessary to manually authorize the chain connector plugin again.
Otherwise, it will not be possible to send CCUs between chains.

> .Authorize sidechain 2

Reviewer: I would use the same name throughout the example, just to be consistent.
```suggestion
.Authorize pos-sidechain-example-two
```

Author: imho it is easier to read like it is
